sled-agent/src/artifact_store.rs

```diff
@@ -0,0 +1,481 @@
+//! direction). This API's implementation is also part of this module.
+//!
+//! POST, PUT, and DELETE operations are handled by the Sled Agent API.
+
+use std::io::ErrorKind;
+use std::net::SocketAddrV6;
+use std::sync::LazyLock;
+use std::time::Duration;
+
+use camino::{Utf8Path, Utf8PathBuf};
+use camino_tempfile::NamedUtf8TempFile;
+use dropshot::{
+    Body, ConfigDropshot, FreeformBody, HttpError, HttpResponseOk,
+    HttpServerStarter, Path, RequestContext, StreamingBody,
+};
+use futures::{Stream, TryStreamExt};
+use illumos_utils::zpool::ZPOOL_MOUNTPOINT_ROOT;
+use omicron_common::address::REPO_DEPOT_PORT;
+use omicron_common::disk::{DatasetKind, DatasetsConfig};
+use omicron_common::update::ArtifactHash;
+use repo_depot_api::*;
+use sha2::{Digest, Sha256};
+use sled_storage::manager::StorageHandle;
+use slog::{info, Logger};
+use tokio::fs::File;
+use tokio::io::AsyncWriteExt;
+
+#[derive(Clone)]
+pub(crate) struct ArtifactStore<T: StorageBackend = StorageHandle> {
+    log: Logger,
+    storage: T,
+}
+
+impl<T: StorageBackend> ArtifactStore<T> {
+    pub(crate) fn new(log: &Logger, storage: T) -> ArtifactStore<T> {
+        ArtifactStore {
+            log: log.new(slog::o!("component" => "ArtifactStore")),
+            storage,
+        }
+    }
+}
+
+impl ArtifactStore {
+    pub(crate) fn start(
+        self,
+        sled_address: SocketAddrV6,
+        dropshot_config: &ConfigDropshot,
+    ) -> Result<
+        dropshot::HttpServer<ArtifactStore<StorageHandle>>,
+        Box<dyn std::error::Error + Send + Sync>,
+    > {
+        let mut depot_address = sled_address;
+        depot_address.set_port(REPO_DEPOT_PORT);
+
+        let log = self.log.new(o!("component" => "dropshot (Repo Depot)"));
+        Ok(HttpServerStarter::new(
+            &ConfigDropshot {
+                bind_address: depot_address.into(),
+                ..dropshot_config.clone()
+            },
+            repo_depot_api_mod::api_description::<RepoDepotImpl>()
+                .expect("registered entrypoints"),
+            self,
+            &log,
+        )?
+        .start())
+    }
+}
+
+impl<T: StorageBackend> ArtifactStore<T> {
+    /// GET operation (served by Repo Depot API)
+    pub(crate) async fn get(
+        &self,
+        sha256: ArtifactHash,
+    ) -> Result<File, Error> {
+        let sha256 = sha256.to_string();
+        for dataset in self.datasets().await? {
+            let path = dataset.join(&sha256);
+            match File::open(&path).await {
+                Ok(file) => {
+                    info!(
+                        &self.log,
+                        "Retrieved artifact";
+                        "sha256" => &sha256,
+                        "path" => path.as_str()
+                    );
+                    return Ok(file);
+                }
+                Err(err) if err.kind() == ErrorKind::NotFound => {}
+                Err(err) => {
+                    return Err(Error::File { verb: "open", path, err });
```

Comment:
Do you think we should keep trying the other datasets in this case, rather than returning early?